net/xia/ppal_ether/main.c

```diff
@@ -18,12 +20,132 @@ static inline struct xip_ether_ctx *ctx_ether(struct xip_ppal_ctx *ctx)
 		: NULL;
 }
 
+/* ETHER-INTERFACE. */
+struct ether_interface {
+	struct net_device  *dev;
+	atomic_t           refcnt;
+	struct rcu_head    rcu_head;
+
+	/* Prevents racing conditions over list. */
+	spinlock_t         interface_lock;
+	struct list_head   list_interface_common_addr;
+};
+
+static inline struct ether_interface *__ether_get_rtnl(
+		const struct net_device *dev)
+{
+	return rtnl_dereference(dev->eth_ptr);
+}
+
+static inline struct ether_interface *ether_interface_get(
+		const struct net_device *dev)
+{
+	struct ether_interface *eint;
+
+	rcu_read_lock();
+	eint = rcu_dereference(dev->eth_ptr);
+	if (eint)
+		atomic_inc(&eint->refcnt);
+	rcu_read_unlock();
+
+	return eint;
+}
+
+static inline void ether_interface_hold(struct ether_interface *eint)
+{
+	atomic_inc(&eint->refcnt);
+}
+
+static inline void ether_interface_put(struct ether_interface *eint)
+{
+	if (atomic_dec_and_test(&eint->refcnt)) {
+		dev_put(eint->dev);
+		kfree(eint);
```

Comment:
Instead of releasing `eint` here, you should call `call_rcu()` here to postpone all the release process, that is, `dev_put()` and `kfree()` must be moved to the RCU callback. This will centralize the release process; right now it's split between here and `eint_destroy()` what can lead to bugs.

Reply:
The only time when it goes inside the { } is when this function is called from inside eint_destroy which is being called by call_rcu. 
If i make the changes, then ether_interface_rcu_put will contain the same code as ether_interface_put now, and ether_interface_put will contain just atomic_dec() as freeing an ether_interface, unless want to, is unnecessary and not possible in the code written.
My suggestion would be to keep the current code. But if you still want me to change the code, I will do so.